File: tests/python/test_createNodeEdges.py
"""
test_createNodeEdges.py

unit test for createNodeEdges().

This function finds kinase pairs that share substrates and creates 2 sheets in workbook:

  - 'tableEdgeSubs': simmetrical matrix NxN. Cell [i,j] contains shared substrates
    between kinase i and kinase j, separated by ';'.

  - 'nodes.edges': list of pairs with columns:
        edge   = "KinasaA.KinasaB"
        weight = number of shared substrates
        subs   = substrates separated by ';'

Run:
    pytest tests/python/test_createNodeEdges.py -v
"""

import pytest
import openpyxl
from ebdtFunctions import createNodeEdges


# ==== TEST DATA ====

# KinaseA y KinaseB share SITE1 and SITE3 : must appear in nodes.edges
# KinaseC only has SITE4 : nothing is shared : doesn`t appear in edges
KINASE_SUBSTRATES = {
    "KinaseA": ["SITE1", "SITE2", "SITE3"],
    "KinaseB": ["SITE1", "SITE3"],
    "KinaseC": ["SITE4"],
}


@pytest.fixture
def empty_wb():
    wb = openpyxl.Workbook()
    wb.remove(wb.active)
    return wb


@pytest.fixture
def wb_con_edges(empty_wb):
    #Runs createNodeEdges and returns the resulting workbook.
    createNodeEdges(empty_wb, KINASE_SUBSTRATES)
    return empty_wb


# ==== TESTS: nodes.edges ====

def test_par_con_sustratos_compartidos_aparece_en_edges(wb_con_edges):
    #KinaseA and KinaseB shere SITE1 and SITE3 : must appear in nodes.edges.
    ws = wb_con_edges["nodes.edges"]
    edges = [row[0] for row in ws.iter_rows(min_row=2, values_only=True) if row[0]]
    assert "KinaseA.KinaseB" in edges


def test_par_sin_sustratos_compartidos_no_aparece_en_edges(wb_con_edges):
    #KinaseA and KinaseC do not share substrates : cannot appear in nodes.edges.
    ws = wb_con_edges["nodes.edges"]
    edges = [row[0] for row in ws.iter_rows(min_row=2, values_only=True) if row[0]]
    assert "KinaseA.KinaseC" not in edges
    assert "KinaseB.KinaseC" not in edges


def test_weight_correcto(wb_con_edges):
    #weight of KinaseA-KinaseB must be 2 (SITE1 and SITE3 are shared).
    ws = wb_con_edges["nodes.edges"]
    for row in ws.iter_rows(min_row=2, values_only=True):
        if row[0] == "KinaseA.KinaseB":
            assert row[1] == 2
            return
    pytest.fail("Cannot find the pair KinaseA.KinaseB in nodes.edges")


def test_sustratos_compartidos_en_edges(wb_con_edges):
    #Column 'subs' must contains shared substrates.
    ws = wb_con_edges["nodes.edges"]
    for row in ws.iter_rows(min_row=2, values_only=True):
        if row[0] == "KinaseA.KinaseB":
            subs = set(row[2].split(";"))
            assert "SITE1" in subs
            assert "SITE3" in subs
            return
    pytest.fail("Cannot find the pair KinaseA.KinaseB in nodes.edges")


# ==== TESTS: tableEdgeSubs ====

def test_tabla_tableEdgeSubs_tiene_dimensiones_correctas(wb_con_edges):
    #tableEdgeSubs must be a matrix: (n_quinasas+1) x (n_quinasas+1).
    ws = wb_con_edges["tableEdgeSubs"]
    rows = list(ws.iter_rows(values_only=True))
    n = len(KINASE_SUBSTRATES)
    assert len(rows) == n + 1      # +1 for the headers row
    assert len(rows[0]) == n + 1   # +1 for the headers column


def test_tabla_tableEdgeSubs_es_simetrica(wb_con_edges):
    """
    tableEdgeSubs[i,j] must be equal to tableEdgeSubs[j,i].
    shared substrates table is simmetrycal by definition.
    """
    ws = wb_con_edges["tableEdgeSubs"]
    rows = list(ws.iter_rows(values_only=True))
    n = len(KINASE_SUBSTRATES)

    for i in range(1, n + 1):
        for j in range(1, n + 1):
            val_ij = rows[i][j] or ""
            val_ji = rows[j][i] or ""
            assert val_ij == val_ji, (
                f"Asymetry in [{i},{j}]: '{val_ij}' != '{val_ji}'"
            )


# ==== TESTS: CREATED SHEETS ====

def test_crea_hoja_tableEdgeSubs(wb_con_edges):
    #Function must create the sheet 'tableEdgeSubs'
    assert "tableEdgeSubs" in wb_con_edges.sheetnames


def test_crea_hoja_nodes_edges(wb_con_edges):
    #Function must create the sheet 'nodes.edges'
    assert "nodes.edges" in wb_con_edges.sheetnames


def test_nodes_edges_tiene_header_correcto(wb_con_edges):
    #sheet nodes.edges must have headers: 'edge', 'weight', 'subs'
    ws = wb_con_edges["nodes.edges"]
    header = list(ws.iter_rows(min_row=1, max_row=1, values_only=True))[0]
    assert header[0] == "edge"
    assert header[1] == "weight"
    assert header[2] == "subs"


#==== TEST: BORDERLINE CASE: no shared substrates ====

def test_sin_sustratos_compartidos_nodes_edges_vacio():
    #if no kinase share substrates, nodes.edges only has a header
    wb = openpyxl.Workbook(); wb.remove(wb.active)
    substrates = {
        "KinaseA": ["SITE1"],
        "KinaseB": ["SITE2"],  #no overlap
    }
    createNodeEdges(wb, substrates)
    ws = wb["nodes.edges"]
    rows = list(ws.iter_rows(min_row=2, values_only=True))
    data_rows = [r for r in rows if any(v is not None for v in r)]
    assert data_rows == []

File: tests/R/test_createNodeEdges.R
# test_createNodeEdges.R

# Unitary tests for createNodeEdges()

# Run:
#   testthat::test_file("tests/R/test_createNodeEdges.R")

source("helpers.R")

# ==== TEST DATA ====

# KinaseA and KinaseB share SITE1 and SITE3 : appear in nodes.edges with weigh 2
# KinaseC only has SITE4 : it shares anything
kinase_substrates <- list(
  KinaseA = c("SITE1", "SITE2", "SITE3"),
  KinaseB = c("SITE1", "SITE3"),
  KinaseC = c("SITE4")
)

wb <- createWorkbook()
addWorksheet(wb, "fold"); addWorksheet(wb, "pvalue")
createNodeEdges(wb, kinase_substrates)


# ==== TESTS: nodes.edges ====

test_that("pair with shared substrates appears in nodes.edges", {
  ws <- readWorkbook(wb, sheet = "nodes.edges", colNames = FALSE)
  edges <- as.character(ws[-1, 1])  #jump header
  expect_true("KinaseA.KinaseB" %in% edges)
})

test_that("pair without shared substrates does not appear in nodes.edges", {
  ws <- readWorkbook(wb, sheet = "nodes.edges", colNames = FALSE)
  edges <- as.character(ws[-1, 1])
  expect_false("KinaseA.KinaseC" %in% edges)
  expect_false("KinaseB.KinaseC" %in% edges)
})

test_that("weight of KinaseA-KinaseB is 2 (SITE1 and SITE3 shared)", {
  ws <- readWorkbook(wb, sheet = "nodes.edges", colNames = FALSE)
  fila <- ws[as.character(ws[, 1]) == "KinaseA.KinaseB", ]
  expect_equal(as.numeric(fila[[2]]), 2)
})

test_that("shared substrates appear in subs column", {
  ws <- readWorkbook(wb, sheet = "nodes.edges", colNames = FALSE)
  fila <- ws[as.character(ws[, 1]) == "KinaseA.KinaseB", ]
  subs <- strsplit(as.character(fila[[3]]), ";")[[1]]
  expect_true("SITE1" %in% subs)
  expect_true("SITE3" %in% subs)
})

test_that("nodes.edges has header: edge, weight, subs", {
  ws <- readWorkbook(wb, sheet = "nodes.edges", colNames = FALSE)
  expect_equal(as.character(ws[1, 1]), "edge")
  expect_equal(as.character(ws[1, 2]), "weight")
  expect_equal(as.character(ws[1, 3]), "subs")
})


# ==== TESTS: tableEdgeSubs ====

test_that("tableEdgeSubs has correct dimensions (n+1 x n+1)", {
  ws <- readWorkbook(wb, sheet = "tableEdgeSubs", colNames = FALSE)
  n <- length(kinase_substrates)
  expect_equal(nrow(ws), n + 1)
  expect_equal(ncol(ws), n + 1)
})

test_that("tableEdgeSubs is simetrycal", {
  ws <- readWorkbook(wb, sheet = "tableEdgeSubs", colNames = FALSE)
  n <- length(kinase_substrates)

  for (i in 2:(n + 1)) {
    for (j in 2:(n + 1)) {
      val_ij <- as.character(ws[i, j] %||% "")
      val_ji <- as.character(ws[j, i] %||% "")
      expect_equal(val_ij, val_ji,
                   label = paste0("asymmetry en [", i, ",", j, "]"))
    }
  }
})


# ==== TESTS: CREATED SHEETS ====

test_that("creates tableEdgeSubs sheet", {
  expect_true("tableEdgeSubs" %in% names(wb))
})

test_that("creates nodes.edges sheet", {
  expect_true("nodes.edges" %in% names(wb))
})


# ==== TEST: BORDERLINE CASE: no shared substrates ====

test_that("nodes.edges only has header if there is no shared substrates", {
  wb2 <- createWorkbook()
  addWorksheet(wb2, "fold"); addWorksheet(wb2, "pvalue")

  substrates_disjuntos <- list(
    KinaseA = c("SITE1"),
    KinaseB = c("SITE2")   #no overlap
  )
  createNodeEdges(wb2, substrates_disjuntos)

  ws <- readWorkbook(wb2, sheet = "nodes.edges", colNames = FALSE)
  #only header row, no data
  data_rows <- ws[-1, ]
  data_rows_con_datos <- data_rows[!is.na(data_rows[, 1]), ]
  expect_equal(nrow(data_rows_con_datos), 0)
})
